sdks/python/apache_beam/runners/trivial_runner.py

```diff
@@ -0,0 +1,415 @@
+    windowing = components.windowing_strategies[
+        components.pcollections[input_pcoll].windowing_strategy_id]
+
+    if (windowing.merge_status ==
+        beam_runner_api_pb2.MergeStatus.Enum.NON_MERGING and
+        windowing.output_time ==
+        beam_runner_api_pb2.OutputTime.Enum.END_OF_WINDOW):
+      # This is the "easy" case, show how to do it by hand.
+      # Note that we're grouping by encoded key, and also by the window.
+      grouped = collections.defaultdict(list)
+      for element in input_elements:
+        for window in element.windows:
+          key, value = element.value
+          grouped[window, key_coder.encode(key)].append(value)
+      output_elements = [
+          windowed_value.WindowedValue(
+              (key_coder.decode(encoded_key), values),
+              window.end, [window],
+              trigger.BatchGlobalTriggerDriver.ONLY_FIRING)
+          for ((window, encoded_key), values) in grouped.items()
+      ]
+    else:
+      # This handles generic merging and triggering.
+      trigger_driver = trigger.create_trigger_driver(
+          execution_state.windowing_strategy(input_pcoll), True)
+      grouped_by_key = collections.defaultdict(list)
+      for element in input_elements:
+        key, value = element.value
+        grouped_by_key[key_coder.encode(key)].append(element.with_value(value))
+      output_elements = []
+      for encoded_key, windowed_values in grouped_by_key.items():
+        for grouping in trigger_driver.process_entire_key(
+            key_coder.decode(encoded_key), windowed_values):
+          output_elements.append(grouping)
+
+    # Store the grouped values in the output PCollection.
+    output_coder = execution_state.windowed_coder(output_pcoll)
+    execution_state.set_pcollection_contents(
+        output_pcoll, [encode_all(output_elements, output_coder)])
+
+  def supported_requirements(self) -> Iterable[str]:
+    # Nothing non-trivial is supported.
+    return []
+
+
+class ExecutionState:
+  """A helper class holding various values and context during execution."""
+  def __init__(self, optimized_pipeline):
+    self.optimized_pipeline = optimized_pipeline
+    self._pcollections_to_encoded_elements = {}
+    self._counter = 0
+    self._process_bundle_descriptors = {}
+    # This emulates a connection to an SDK worker (e.g. its data, control,
+    # etc. channels).
+    # There are other variants available as well (e.g. GRPC, Docker, ...)
+    self.worker_handler = worker_handlers.EmbeddedWorkerHandler(
+        None,
+        state=worker_handlers.StateServicer(),
+        provision_info=None,
+        worker_manager=self)
+    self._windowed_coders = {}
+    # Populate all windowed coders before creating _pipeline_context.
+    for pcoll_id in self.optimized_pipeline.components.pcollections.keys():
+      self.windowed_coder_id(pcoll_id)
+    self._pipeline_context = pipeline_context.PipelineContext(
+        optimized_pipeline.components)
```

Comment:
`PipelineContext` is marked as an internal only class

Reply:
Fair point. Runner implementations sit kind-of on the boundary needing some internal visibility (though it'd be nice to minimize this). I think this is OK (it's stable for one).